### DISCUSSION TITLE
Help Subscribing to Probe Status Service by MAC Address

### DISCUSSION THREAD

**Discussion by flynace** [h2zero/NimBLE-Arduino#667] [Q&A]
This is my first time trying to use NimBLE and an ESP32-S3.
I have a few probes and I know their MAC addresses and their Probe Status Service (UUID 00000100-CAAB-3792-3D44-97AE51C1407A) which has a single Service Characteristic (UUID 00000101-CAAB-3792-3D44-97AE51C1407A) that supports BLE notifications.

When I run the NimBLE_Client example sketch it sees a probe...
**_Advertised Device: Name: , Address: d2:71:05:5a:a6:69, manufacturer data: c7090116230010800300000000000000000000000dc00000, serviceUUID: 0xfe59_**

I tried using a Whitelist example from another discussion, but it doesn't connect based on the MAC address.

In the NimBLE_Client sketch, if I change the "DEAD" to "fe59" in the line...
`pSvc = pClient->getService("fe59");`
...it does connect, but the example if so complicated I cannot figure out how to complete the subscription to the notification.
I do not know what the fe59 service is, but it shows up in the advertisement of these probes.
And I want to whitelist based on a single MAC.

I previously could get it to work in the ArduinoBLE library if I roughly did...
```
BLE.scanForAddress("d2:71:05:5a:a6:69");
peripheral.connect();
peripheral.discoverAttributes();
BLEService service = peripheral.service("00000100-caab-3792-3d44-97ae51c1407a");
BLECharacteristic characteristic = service.characteristic("00000101-caab-3792-3d44-97ae51c1407a");
characteristic.subscribe();
```
...but it wasn't great, and I really want to use NimBLE on the ESP32.
Is there an easy conversion of this to NimBLE?
ArduinoBLE had to do the _Discover Attributes_ even though they are already known, but I can't seem to find a NimBLE equivalent.

I also need to run this in the main loop so if the connection gets dropped it will just keep retrying.

Thank you in advance for any guidance.

**Comment by flynace** [comment]
Not sure why my first search didn't uncover it, but I found `NimBLEClient::discoverAttributes()` 
Now I just need to next figure out how to use it...

**Reply by h2zero** [reply]
Take a look at some of the simpler examples. If you know the service and characteristic UUID you can just get those instead of doing a full discovery.

**Reply by flynace** [reply]
Thank you for the advice @h2zero 
If I run the NimBLE_Scan_Continuous example it sees the probe and prints every time.
But when I try to add the Whitelist feature the callback never prints.
I believe if I can get the Whitelist to work I will be able to figure it out.
Can you advise why the following code is not working?
```
#include "NimBLEDevice.h"

NimBLEScan* pBLEScan;
NimBLEAddress a("d2:71:05:5a:a6:69");

class MyAdvertisedDeviceCallbacks : public NimBLEAdvertisedDeviceCallbacks {
  void onResult(NimBLEAdvertisedDevice* advertisedDevice) {
    Serial.println(advertisedDevice->getAddress().toString().c_str());
  }
};

void setup() {
  while (!Serial && millis() < 5000);
  Serial.begin(9600);
  Serial.println("Starting...");

  NimBLEDevice::setScanFilterMode(CONFIG_BTDM_SCAN_DUPL_TYPE_DEVICE);
  NimBLEDevice::setScanDuplicateCacheSize(200);

  NimBLEDevice::init("");
  NimBLEDevice::whiteListAdd(a);

  pBLEScan = NimBLEDevice::getScan();  //create new scan
  pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks(), true);
  pBLEScan->setFilterPolicy(BLE_HCI_SCAN_FILT_USE_WL);
  pBLEScan->setActiveScan(true);  // Set active scanning, this will get more data from the advertiser.
  pBLEScan->setInterval(97);      // How often the scan occurs / switches channels; in milliseconds,
  pBLEScan->setWindow(37);        // How long to scan during the interval; in milliseconds.
  pBLEScan->setMaxResults(0);     // do not store the scan results, use callback only.
}

void loop() {
  // If an error occurs that stops the scan, it will be restarted here.
  if (pBLEScan->isScanning() == false) {
    // Start scan with: duration = 0 seconds(forever), no scan end callback, not a continuation of a previous scan.
    pBLEScan->start(0, nullptr, false);
  }

  delay(2000);
}
```
Thank you!

**Reply by h2zero** [reply]
You need to set the address type to the same as the device advertises with. try: `NimBLEAddress a("d2:71:05:5a:a6:69", 1);` As it is likely a random static address.
